Lib/_pyrepl/readline.py

```diff
@@ -596,3 +610,321 @@ def _setup(namespace: Mapping[str, Any]) -> None:
+
+        if not (module := self.import_module(path)):
+            return []
+
+        submodules = self.filter_submodules(module, prefix)
+        if not is_relative:
+            return [f'{path}.{name}' for name in submodules]
+        return [f'.{name}' for name in submodules]
+
+    def import_module(self, path: str) -> ModuleType | None:
+        package = self.namespace.get('__package__')
+        is_relative = path.startswith('.')
+        if is_relative and not package:
+            return None
+        try:
+            module = importlib.import_module(
+                path,
+                package=package if is_relative else None)
+        except ImportError:
+            return None
+        return module
+
+    def filter_submodules(self, module: ModuleType, prefix: str) -> list[str]:
+        if not hasattr(module, '__path__'):
+            return []
+        return [name for _, name, _ in pkgutil.iter_modules(module.__path__)
+                if name.startswith(prefix)]
+
+    def filter_attributes(self, module: ModuleType, prefix: str) -> list[str]:
+        return [attr for attr in module.__dict__ if attr.startswith(prefix)]
+
+    def get_path_and_prefix(self, dotted_name: str) -> tuple[str, str]:
+        if '.' not in dotted_name:
+            return '', dotted_name
+        if dotted_name.startswith('.'):
+            stripped = dotted_name.lstrip('.')
+            dots = '.' * (len(dotted_name) - len(stripped))
+            if '.' not in stripped:
+                return dots, stripped
+            path, prefix = stripped.rsplit('.', 1)
+            return dots + path, prefix
+        path, prefix = dotted_name.rsplit('.', 1)
+        return path, prefix
+
+    @property
+    def global_cache(self) -> list[str]:
+        if not self._global_cache or self._curr_sys_path != sys.path:
+            self._curr_sys_path = sys.path[:]
+            self._global_cache = [
+                name for _, name, _ in pkgutil.iter_modules()]
+        return self._global_cache
+
+
+class ImportParser:
+    """
+    Parses incomplete import statements that are
+    suitable for autocomplete suggestions.
+
+    Examples:
+        - import foo          -> Result(from_name=None, name='foo')
+        - import foo.         -> Result(from_name=None, name='foo.')
+        - from foo            -> Result(from_name='foo', name=None)
+        - from foo import bar -> Result(from_name='foo', name='bar')
+        - from .foo import (  -> Result(from_name='.foo', name='')
+
+    Note that the parser works in reverse order, starting from the
+    last token in the input string. This makes the parser more robust
+    when parsing multiple statements.
+    """
+    _ignored_tokens = {
+        tokenize.INDENT, tokenize.DEDENT, tokenize.COMMENT,
+        tokenize.NL, tokenize.NEWLINE, tokenize.ENDMARKER
+    }
+    _keywords = {'import', 'from', 'as'}
+
+    def __init__(self, code: str):
```

Comment:
We don't have mypy configured (yet?) to enforce return annotations for this module, but let's add them.
```suggestion
    def __init__(self, code: str) -> None:
```